### ISSUE TITLE
Client metrics collection should be opt-in, not opt-out

### ISSUE THREAD

**Issue by asarkar** [yidongnan/grpc-spring-boot-starter#398]
**The context**

Remove unnecessary clutter from application logs.

**The bug**

When the gRPC server is not available, closing the client call throws an exception.

**Stacktrace and logs**
```
2020-07-31 23:48:41.208 [grpc-default-executor-1] [ERROR] i.g.i.SerializingExecutor - Exception while executing runnable io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed@36bb90cc
io.grpc.StatusRuntimeException: INTERNAL
	at io.grpc.Status.asRuntimeException(Status.java:524)
	at mycompany.myservice.ResponseHandler.onError(ResponseHandler.kt:117)
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:460)
	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at net.devh.boot.grpc.client.metric.MetricCollectingClientCallListener.onClose(MetricCollectingClientCallListener.java:65)
```

Somewhere deep in the stacktrace is the root cause:
```
Caused by: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: bad-server/x.x.x.x:123
```

**Steps to Reproduce**

N/A

**The application's environment**

Which versions do you use?
* Spring (boot): 2.3.1.RELEASE
* grpc-java: 1.30.2
* grpc-spring-boot-starter: 2.9.0.RELEASE

I see two problem here:

1. Client metrics collection should be opt-in, not opt-out; for now, excluding [GrpcClientMetricAutoConfiguration](https://github.com/yidongnan/grpc-spring-boot-starter/blob/master/grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/autoconfigure/GrpcClientMetricAutoConfiguration.java) in `application.yml` reduces the chatter in the logs. Note that Spring also has various interceptors that'll log this, so will gRPC. In the end, we end up with a stacktrace that's as long as a Sunday newspaper.
2. `MetricCollectingClientCallListener.onClose` should probably log at WARN level, and most importantly, without the stacktrace, that the call couldn't be closed. Even better, if there is a logging detail configuration parameter, like `error-detail: ALL, FIRST, MESSAGE, NONE`. `ALL` is current behavior; `FIRST` shows the first stacktrace; `MESSAGE` shows the error message only, and `NONE` suppresses the exception. Another option is `error-detail=0,1,...,Integer.MAX_VALUE`

**Comment by ST-DDT** [comment]
> Remove unnecessary clutter from application logs.

If you are referring to the long stacktraces there is hardly anything I can do about that. Please report that issue directly to grpc-java.

If you think that the metric interceptor logs that exception, then you are mistaken:

https://github.com/yidongnan/grpc-spring-boot-starter/blob/55981c9632ec2de6ebd882df08fcec1c2c277533/grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/metric/MetricCollectingClientCallListener.java#L62-66

This is just grpcs way of calling things. In spring web you can observe the same behavior with servlet filters.

Your comments sound like excluding the metric interceptors reduces the logs by a large amount.
Could you please attach the full error log/repeated stacktraces so I can analyze the behavior in more detail? If possible for both variants (metrics enabled and disabled) in order to compare them.

Metric collection in spring-boot is opt-out and not opt-in. In order to be consistent with the base library, I prefer to stay the same. This library only collects metrics if the metrics system is enabled for the entire application. We explicitly moved the metrics configuration to a separate class in order to allow them to be turned off selectively.

If you really aren't interested in metric collectionbat all, then you should disable it entirely e.g. by excluding `org.springframework.boot.actuate.autoconfigure.metrics.CompositeMeterRegistryAutoConfiguration` (not sure whether that is the correct way to do that).

**Comment by asarkar** [comment]
> Metric collection in spring-boot is opt-out and not opt-in

That's because Boot was created with the implicit motto "_web application for monkeys_", heavily influenced by Grails' strong assumptions about what the user wants. I should know, I've been using it since before the first release, and am also a committer. There's no reason that your library has to adhere to the same principles or make the same mistakes, simply because you provide additional functionality on top of Boot. Many times, Boot autoconfigs come in the way of advanced usage, and negatively impact performance; see [this SO discussion](https://stackoverflow.com/a/49663075/839733).

> Could you please attach the full error log/repeated stacktraces

I ran the same integration test with several configuration tweaks:

1. [default.txt](https://github.com/yidongnan/grpc-spring-boot-starter/files/5011501/default.txt); OOTB, total stacktrace: 273 lines.
2. [no-metrics.txt](https://github.com/yidongnan/grpc-spring-boot-starter/files/5011500/no-metrics.txt):
```
    autoconfigure.exclude: >-
    org.springframework.boot.actuate.autoconfigure.metrics.web.reactive.WebFluxMetricsAutoConfiguration,
    net.devh.boot.grpc.client.autoconfigure.GrpcClientMetricAutoConfiguration
```
total stacktrace: 255 lines.
3. [no-metrics-no-chatter.txt](https://github.com/yidongnan/grpc-spring-boot-starter/files/5011499/no-metrics-no-chatter.txt): Same as 2 +
```
org.springframework.web.HttpLogging, org.springframework.boot.autoconfigure.web.reactive.error
```
packages logging turned off
total stacktrace: 218 lines.
4.[no-autoconfig.txt](https://github.com/yidongnan/grpc-spring-boot-starter/files/5011498/no-autoconfig.txt): 
```
autoconfigure.exclude: org.springframework.boot.actuate.autoconfigure.metrics.CompositeMeterRegistryAutoConfiguration
```
total stacktrace: 273 lines.

All stacktraces include JUnit.

Clearly, `#3` has the minimal noise. More reduction is possible by cutting on gRPC logs, but then I lose the cause of the error, and hence the origin in the source code.

**Comment by ST-DDT** [comment]
This is really a tough one for me.

As for the issue with the long stacktraces, its just 4 lines. I can reduce it to 3 lines, but that doesn't matter in comparison to 273 lines or just 218. AFAICT logging the stacktrace only once instead of twice or thrice would do much better than that. The actual error message is 34 lines and the other 200 lines are the noise you should try to get rid of.
If you are able to share some of the relevant code to reproduce the above stacktrace, I'm willing to analyze it in order to reduce the total number of stacktraces. As for the 3-4 lines in the stacktrace, I consider it negligible.

As for the performance, I agree with your that spring-boots (startup) performance is really bad. However, the SO question that you linked also had a easy solution to it: Explicitly list auto configurations to load.
If you really after each tiny piece of performance, than you will definitely turn off every feature you don't need including the `MeterRegistry`. If you have turned it off, this library won't add the metric collectors either.
Aside from the small startup impact and the 4 lines of stacktrace, the runtime impact is really tiny.

I also talked to a good friend of mine about it and we came to the same conclusion:
It is easier to turn off a function you know and don't need than to turn on a function you need and don't know.

I won't turn the metrics off by default.

I leave this issue open in case you wish to discuss the reduction of the stacktrace repetition in your logs.

**Comment by asarkar** [comment]
> It is easier to turn off a function you know and don't need than to turn on a function you need and don't know.

That's like forcing me to buy things I don't need, and suggesting that I start a return process. Also, the assumption that a user knows about this magic autoconfig class hidden inside a JAR is not based on reality.

Let's agree to disagree; at the end of the day, you're the maintainer of the library, and can choose to do whatever you want with it. I made my case, and if I don't like your response, I'm free to not use the library. That's how democracy works.

Regarding the length of the stacktrace, I can't post here proprietary code, and it'll take me some time to make a minimal, reproducible example. If you are ok leaving this thread open, that's fine, and I'll post updates when I can; if not, feel free to close it.

Last but not the least, last I checked, Spring Boot was having some issues with Graal. Perhaps they have solved those issues since, but I didn't go back and check.

Thanks.

**Comment by YixingCheng** [comment]
>Let's agree to disagree; at the end of the day, you're the maintainer of the library, and can choose to do whatever you want with it. I made my case, and if I don't like your response, I'm free to not use the library. That's how democracy works.

It is totally disgusting to try to teach or educate others what is democrazy. It is like only you know what is democrazy and the maintainer is being a dictator here. Drop your white supremacy and show some respect.

**Comment by asarkar** [comment]
@YixingCheng Go back into your cave, your comment has been reported. Deleting it would be waste of time, I got a screenshot.

**Comment by ST-DDT** [comment]
Closing it for now. Feel free to open a new issue when you have a reproducible example.
